pkg/kubelet/qos/policy.go

```diff
@@ -63,12 +63,33 @@ func GetContainerOOMScoreAdjust(pod *v1.Pod, container *v1.Container, memoryCapa
 	// targets for OOM kills.
 	// Note that this is a heuristic, it won't work if a container has many small processes.
 	memoryRequest := container.Resources.Requests.Memory().Value()
+
 	if utilfeature.DefaultFeatureGate.Enabled(features.InPlacePodVerticalScaling) {
 		if cs, ok := podutil.GetContainerStatus(pod.Status.ContainerStatuses, container.Name); ok {
 			memoryRequest = cs.AllocatedResources.Memory().Value()
 		}
 	}
 	oomScoreAdjust := 1000 - (1000*memoryRequest)/memoryCapacity
+
+	// adapt the sidecarContainer memoryRequest for OOM ADJ calculation
+	// calculate the oom score adjustment based on: min-memory( currentSideCarContainer , min-memory(regular containers) ) .
+	if utilfeature.DefaultFeatureGate.Enabled(features.SidecarContainers) && isSidecarContainer(pod, container) {
+		// check min memory quantity in regular containers
+		minMemoryRequest := minRegularContainerMemory(*pod)
+		// check min memory quantity with the current sidecarContainer
+		// inPlacePodVerticalScaling is not possible for the moment in InitContainers
+		if minMemoryRequest > container.Resources.Requests.Memory().Value() {
```

Comment:
is it accurate? I think we need to take `minMemoryRequest` if is it greater than the sidecar's memory. We want to keep sidecar's oom score if it is better than any of containers inside the Pod. And if it is worse than all containers inside the pod, take the minimum of containers

Reply:
thanks for your feedback . 
```go
		// check min memory quantity in regular containers
		minMemoryRequest := minRegularContainerMemory(*pod)
		// check min memory quantity with the current sidecarContainer
		// inPlacePodVerticalScaling is not possible for the moment in InitContainers
		if minMemoryRequest > container.Resources.Requests.Memory().Value() {
			minMemoryRequest = container.Resources.Requests.Memory().Value()
		}
```

in the first step I calculate the minimum memory in all regular containers then I check this memory with the current container (sidecar)  : min ( memory all regular containers,   memory sidecar container)  .

Normally I use the minimum.

Reply:
but this logic is wrong. You need to take max out of minimal container memory and the sidecars memory. This will give you the best score.

The logic we want is - sidecar's OOM score is NO WORSE then the WORST score of all containers.

Reply:
> You need to take max out of minimal container memory and the sidecars memory. This will give you the best score.

as a result the behavior on big sidecar container will be : 

   sidecar :   400Mi
   main-1 :   200Mi
   main-2:    100Mi

**max** out of (minimal container memory and the sidecars memory): 400Mi   
    score adj **sidecar**  **<**  score adj  main  
( oomScoreAdjust = 1000 - (1000*memoryRequest)/memoryCapacity )

> The logic we want is - sidecar's OOM score is NO WORSE then the WORST score of all containers.  

I will remove this : 
```go
		// when OOM score adj currentSideCarContainer <= oom score adj min-memory(regular containers)
		// ==> add 1 to the sidecar container oom score adj (first to kill).
		if minMemoryRequest <= container.Resources.Requests.Memory().Value() {
			oomScoreAdjust += 1
		}
```

Reply:
So the end goal is to make sidecar's behavior as the following:

1. If probability of OOM-killing sidecar is less then at least one regular containers, keep the old formulae
2. If probability of OOM-killing is higher then all regular containers, use the same probability as the worst container from regular containers

Reply:
I feel like we are talking past each other. Maybe some of my understandings are wrong.

My understandings are: 

1. The OOMscore - the lower value is better. 
2. The OOMscore is calculated based on memory compared to total memory on the node. Higher the memory request of a container - lower the chance it will be OOM killed.

For sidecars we have an issue that typically sidecar's memory is the lowest of all. So it is OOM killed fastest. We want to adjust it so it is not an outlier.

So we want to calculate the minimum of all containers oom scores (`minMemoryOomScoreAdjust`). And if the value is lower (better) than the `oomScoreAdjust`, than set the `oomScoreAdjust` to it. Opposite of what you are doing here:

```
		if oomScoreAdjust < minMemoryOomScoreAdjust {
			oomScoreAdjust = minMemoryOomScoreAdjust
		}
```

Am I missing anything? Please let me know if my understanding is wrong here.

Reply:
> 1. The OOMscore - the lower value is better.
> 2. The OOMscore is calculated based on memory compared to total memory on the node. Higher the memory request of a container - lower the chance it will be OOM killed.

Got it. I initially thought the sidecar container would be the first to be killed. However, your proposal better aligns with the importance of keeping the sidecar running, prioritizing it to be killed only at a final stage.

Updated the PR and [kep ](https://github.com/kubernetes/enhancements/pull/4932).  

Thank you.